services/src/datasets/external/netcdfcf/mod.rs

```diff
@@ -0,0 +1,1109 @@
+
+        unsafe {
+            let c_group_names = GDALGroupGetGroupNames(self.c_group, options.as_ptr());
+
+            Self::_string_array(c_group_names)
+        }
+    }
+
+    fn open_group(&self, name: &str) -> Result<Self> {
+        let c_name = CString::new(name).expect("no null-byte in name");
+        let options = CslStringList::new();
+
+        let c_group = unsafe {
+            let c_group = GDALGroupOpenGroup(self.c_group, c_name.as_ptr(), options.as_ptr());
+
+            if c_group.is_null() {
+                return Err(NetCdfCf4DProviderError::GdalMd {
+                    source: Self::_last_null_pointer_err("GDALGroupOpenGroup"),
+                });
+            }
+
+            c_group
+        };
+
+        #[allow(clippy::used_underscore_binding)]
+        Ok(Self {
+            c_group,
+            _dataset: self._dataset,
+            name: name.to_string(),
+        })
+    }
+
+    fn datatype_of_numeric_array(&self, name: &str) -> Result<RasterDataType> {
+        let name = CString::new(name).expect("no null-byte in name");
+
+        let gdal_data_type = unsafe {
+            let options = CslStringList::new();
+            let c_mdarray = GDALGroupOpenMDArray(self.c_group, name.as_ptr(), options.as_ptr());
+
+            if c_mdarray.is_null() {
+                return Err(NetCdfCf4DProviderError::GdalMd {
+                    source: Self::_last_null_pointer_err("GDALGroupOpenMDArray"),
+                });
+            }
+
+            let c_data_type = GDALMDArrayGetDataType(c_mdarray);
+
+            GDALExtendedDataTypeGetNumericDataType(c_data_type)
+        };
+
+        Ok(match gdal_data_type {
+            GDALDataType::GDT_Byte => RasterDataType::U8,
+            GDALDataType::GDT_UInt16 => RasterDataType::U16,
+            GDALDataType::GDT_Int16 => RasterDataType::I16,
+            GDALDataType::GDT_UInt32 => RasterDataType::U32,
+            GDALDataType::GDT_Int32 => RasterDataType::I32,
+            GDALDataType::GDT_Float32 => RasterDataType::F32,
+            GDALDataType::GDT_Float64 => RasterDataType::F64,
+            _ => {
+                return Err(NetCdfCf4DProviderError::UnknownGdalDatatype {
+                    type_number: gdal_data_type,
+                })
+            }
+        })
+    }
+
+    fn open_array(&self, name: &str) -> Result<MdArray> {
+        let name = CString::new(name).expect("no null-byte in name");
+        let options = CslStringList::new();
+
+        Ok(unsafe {
+            let c_mdarray = GDALGroupOpenMDArray(self.c_group, name.as_ptr(), options.as_ptr());
+
+            if c_mdarray.is_null() {
+                return Err(NetCdfCf4DProviderError::GdalMd {
+                    source: Self::_last_null_pointer_err("GDALGroupOpenMDArray"),
+                });
+            }
+
+            MdArray {
+                c_mdarray,
+                _group: self,
+            }
+        })
+    }
+}
+/// TODO: There are potential memory leaks from the C-API
```

Comment:
we should not merge code where we introduce memory leaks

Reply:
s.o.

Reply:
I removed all leaks and checked the tests with ASAN:

`RUSTFLAGS="-Z sanitizer=leak" cargo test --package geoengine-services --lib --all-features --target x86_64-unknown-linux-gnu -- datasets::external::netcdfcf::tests`

e109efe